<a href="https://colab.research.google.com/github/Art207141/Lab_2_sbor_analyz/blob/main/lab02/Lab02.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Лабораторная работа 02
### Часть 1. Базовый конвейер анализа данных в Pandas
Иванов Артём Иванович ЕТ-216

# 3. Формирование и сохранение исходных данных

In [1]:
# 3.1
import numpy as np
import pandas as pd
raw_data = {
"order_id": [
1001, 1002, 1003, 1004, 1005,
1006, 1007, 1008, 1009, 1010,
1011, 1012, 1013, 1014, 1015
],
"channel": [
"site", "app", "app", "marketplace", "site",
"app", "site", "marketplace", "app", "site",
"marketplace", "app", "site", "site", "marketplace"
],
"order_sum": [
1200, 2500, 1800, 3200, 900,
1500, 2800, 45000, 2100, 1700,
3600, 1100, 2400, 1900, 3300
],
"delivery_time": [
25, np.nan, 40, 55, 20,
30, np.nan, 70, 35, 45,
50, 18, 27, np.nan, 60
],
"rating": [
5, 4, 5, 3, 5,
4, 2, 1, 5, 4,
3, 5, 4, 5, 2
]
}
df = pd.DataFrame(raw_data)
print(df)
df.to_csv("online_store_orders.csv", index=False)

    order_id      channel  order_sum  delivery_time  rating
0       1001         site       1200           25.0       5
1       1002          app       2500            NaN       4
2       1003          app       1800           40.0       5
3       1004  marketplace       3200           55.0       3
4       1005         site        900           20.0       5
5       1006          app       1500           30.0       4
6       1007         site       2800            NaN       2
7       1008  marketplace      45000           70.0       1
8       1009          app       2100           35.0       5
9       1010         site       1700           45.0       4
10      1011  marketplace       3600           50.0       3
11      1012          app       1100           18.0       5
12      1013         site       2400           27.0       4
13      1014         site       1900            NaN       5
14      1015  marketplace       3300           60.0       2


## Этап 3.1 —  Формирование и сохранение исходных данных

На базе предоставленного словаря `raw_data` мы успешно создали объект DataFrame с именем `df`.
В получившейся матрице зафиксировано 15 строк (заказов) и 5 колонок с ключевыми метриками (уникальный номер, канал привлечения, итоговый чек, длительность логистики и оценка пользователя). Табличная часть экспортирована в документ `online_store_orders.csv`.

**Разбор контрольного вопроса:**  
Инструкция `index=False` исключает автоматическую нумерацию строк Pandas из структуры сохраняемого файла. Если проигнорировать это правило, то при повторном импорте данных через `read_csv` приложение сгенерирует технический безымянный столбец, что приведет к дублированию индексов и загрязнению датасета.


# 4. Загрузка и первичный анализ данных

In [2]:
# 4.1
store_df = pd.read_csv("online_store_orders.csv")
print("Первые 5 строк:")
print(store_df.head())

print("\nПоследние 5 строк:")
print(store_df.tail())

print("\nРазмер таблицы:")
print(store_df.shape)

print("\nНазвания столбцов:")
print(store_df.columns)

print("\nИнформация о таблице:")
print(store_df.info())

Первые 5 строк:
   order_id      channel  order_sum  delivery_time  rating
0      1001         site       1200           25.0       5
1      1002          app       2500            NaN       4
2      1003          app       1800           40.0       5
3      1004  marketplace       3200           55.0       3
4      1005         site        900           20.0       5

Последние 5 строк:
    order_id      channel  order_sum  delivery_time  rating
10      1011  marketplace       3600           50.0       3
11      1012          app       1100           18.0       5
12      1013         site       2400           27.0       4
13      1014         site       1900            NaN       5
14      1015  marketplace       3300           60.0       2

Размер таблицы:
(15, 5)

Названия столбцов:
Index(['order_id', 'channel', 'order_sum', 'delivery_time', 'rating'], dtype='object')

Информация о таблице:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15 entries, 0 to 14
Data columns (total 5 col

## Этап 4.1 — Чтение файла и базовый аудит структуры

Данные из сохраненного CSV-файла импортированы в новый DataFrame `store_df`.
В ходе первичного осмотра были выведены граничные строки (первая и последняя пятерки), а также зафиксированы габариты таблицы — 15 наблюдений на 5 признаков.

**Ответы на контрольные вопросы:**
1. Общее число записей (строк): 15
2. Общее количество признаков (колонок): 5
3. Колонки с числовым форматом: order_id, order_sum, delivery_time, rating
4. Переменные, содержащие пропуски: delivery_time
5. Итоговое количество пустых ячеек: 3

*Примечание:* Для метрики delivery_time среда автоматически установила тип float64. Это обусловлено присутствием пустых значений (NaN), которые в стандартной библиотеке Pandas не могут быть обработаны внутри целочисленного формата int64.


# 5. Выборка и фильтрация данных

In [3]:
# 5.1
selected_df = store_df[["order_sum", "delivery_time"]]
print(selected_df)

    order_sum  delivery_time
0        1200           25.0
1        2500            NaN
2        1800           40.0
3        3200           55.0
4         900           20.0
5        1500           30.0
6        2800            NaN
7       45000           70.0
8        2100           35.0
9        1700           45.0
10       3600           50.0
11       1100           18.0
12       2400           27.0
13       1900            NaN
14       3300           60.0


In [4]:
# 5.2
slice_df = store_df.iloc[3:8, [2, 3]]
print(slice_df)

   order_sum  delivery_time
3       3200           55.0
4        900           20.0
5       1500           30.0
6       2800            NaN
7      45000           70.0


## Этапы 5.1 и 5.2 — Выбор столбцов и строк

5.1: С помощью синтаксиса `store_df[["order_sum", "delivery_time"]]` выполнено изолирование двух целевых колонок (стоимость и время доставки) в отдельную структуру.

5.2: Инструмент `.iloc[3:8, ]` позволил извлечь срез, состоящий из строк с индексами от 3 до 7 включительно, ограничив выборку только показателями чека и времени. Индекс 8 указан в качестве правой границы, поскольку в концепции работы `.iloc` крайнее правое значение исключается из области поиска.


In [5]:
# 5.3
premium_orders = store_df[(store_df["order_sum"] > 2000) & (store_df["rating"] == 5)]
print(premium_orders)

   order_id channel  order_sum  delivery_time  rating
8      1009     app       2100           35.0       5


## Этап 5.3 — Фильтрация данных

Применение логического фильтра `(store_df["order_sum"] > 2000) & (store_df["rating"] == 5)` выявило ровно одно совпадение в общей базе. Данным критериям соответствует только покупка под номером 1009, совершенная через мобильное приложение (сумма составила 2100 единиц, доставка заняла 35 минут, а покупатель поставил высший балл).

**Разбор контрольного вопроса:**  
Использование побитового знака `&` вместо текстового союза `and` продиктовано спецификой Pandas. Оператор `and` ориентирован на оценку единичных логических переменных и вызывает сбой при попытке обработать массив данных. Значок `&` осуществляет векторизованную (построчную) проверку всего столбца, формируя корректную булеву маску.


# 6. Проверка качества данных и статистический анализ

In [6]:
# 6.1
print(store_df.isna().sum())

order_id         0
channel          0
order_sum        0
delivery_time    3
rating           0
dtype: int64


## Этап 6.1 — Пропущенные значения

Команда `store_df.isna().sum()` позволила рассчитать плотность заполнения таблицы. Проблемная зона обнаружена исключительно в метрике времени доставки (`delivery_time`), где отсутствует 3 значения. Все остальные параметры заполнены на 100%.

**Резюме:** Три клиентские транзакции не содержат сведений о сроках выполнения доставки.


In [7]:
# 6.2
print(store_df.describe())

          order_id    order_sum  delivery_time     rating
count    15.000000     15.00000      12.000000  15.000000
mean   1008.000000   5000.00000      39.583333   3.800000
std       4.472136  11096.33144      16.654010   1.320173
min    1001.000000    900.00000      18.000000   1.000000
25%    1004.500000   1600.00000      26.500000   3.000000
50%    1008.000000   2100.00000      37.500000   4.000000
75%    1011.500000   3000.00000      51.250000   5.000000
max    1015.000000  45000.00000      70.000000   5.000000


## Этап 6.2 — Описательная статистика

Функция `store_df.describe()` сформировала математический профиль числовых колонок:

- По чекам (`order_sum`): обработано 15 позиций, средний показатель равен 5000, стандартное отклонение составляет 11096, центральное значение (медиана) — 2100, пиковое значение — 45000.
- По логистике (`delivery_time`): учтено 12 заполненных строк, среднее время равно 39.58 мин., медиана зафиксирована на отметке 37.5 мин., максимум — 70 мин.
- По оценкам (`rating`): собрано 15 отзывов со средним баллом 3.8 в границах от 1 до 5.

**Разбор контрольного вопроса:**  
В столбце `order_sum` наблюдается экстремальный максимум на уровне 45 000 рублей, который резко выбивается из стандартного диапазона трат (остальные транзакции колеблются от 900 до 3 600 рублей). Эта аномалия сильно смещает среднее арифметическое (5000) вверх, делая его нерепрезентативным, в то время как медиана (2100) сохраняет объективность.


In [8]:
# 6.3
mean_delivery = store_df["delivery_time"].mean()
median_delivery = store_df["delivery_time"].median()

print("Среднее:", mean_delivery)
print("Медиана:", median_delivery)

Среднее: 39.583333333333336
Медиана: 37.5


## Этап 6.3 — Среднее и медиана

Показатель среднего времени транспортировки составил 39.58 мин., а медианный маркер зафиксирован на уровне 37.5 мин. Инструменты вычислений Pandas автоматически исключили пустые ячейки (NaN) из расчетов.

**Разбор контрольного вопроса:**  
Расхождение между этими метриками обусловлено разной математической природой алгоритмов. Среднее значение учитывает вес абсолютно каждого элемента и легко поддается влиянию аномалий. Медиана отражает позицию центрального элемента в упорядоченном ряду, оставаясь невосприимчивой к крайним значениям. В нашей выборке присутствует долгая доставка (70 минут), которая сдвинула среднее значение чуть выше медианы.


In [9]:
# 6.4
print(store_df["rating"].value_counts())

rating
5    6
4    4
3    2
2    2
1    1
Name: count, dtype: int64


## Этап 6.4 — Распределение рейтингов

Метод `value_counts()` сгруппировал оценки пользователей по частоте их появления:
- Высший балл (5) выставлен 6 раз;
- Хорошо (4) встречается у 4 заказов;
- Удовлетворительно (3) и посредственно (2) набрали по 2 упоминания;
- Минимальный балл (1) зафиксирован единожды.

**Разбор контрольного вопроса:**  
Наиболее популярным откликом среди аудитории является оценка «5» (6 упоминаний). Доминирование максимального балла свидетельствует о высоком уровне удовлетворенности сервисом у преобладающей части клиентов.


# 7. Группировка и получение аналитического результата

In [10]:
# 7.1
store_df["is_high_value"] = store_df["order_sum"] > 3000
print(store_df[["order_id", "order_sum", "is_high_value"]])

    order_id  order_sum  is_high_value
0       1001       1200          False
1       1002       2500          False
2       1003       1800          False
3       1004       3200           True
4       1005        900          False
5       1006       1500          False
6       1007       2800          False
7       1008      45000           True
8       1009       2100          False
9       1010       1700          False
10      1011       3600           True
11      1012       1100          False
12      1013       2400          False
13      1014       1900          False
14      1015       3300           True


## Этап 7.1 — Создание производного признака

С помощью конструкции `store_df["order_sum"] > 3000` в структуру таблицы внедрен новый бинарный признак `is_high_value`. Маркер принимает значение `True`, если затраты на заказ превысили планку в 3000 рублей, и `False` во всех остальных случаях.

Под критерий дорогих покупок подпали 4 транзакции (идентификаторы: 1004, 1008, 1011, 1015).


In [11]:
# 7.2
channel_stats = store_df.groupby("channel")["order_sum"].mean()
print(channel_stats)

channel
app             1800.000000
marketplace    13775.000000
site            1816.666667
Name: order_sum, dtype: float64


## Этап 7.2 — Группировка по каналу продаж

Операция `groupby("channel")["order_sum"].mean()` позволила рассчитать средние чеки для каждого источника продаж:
- Мобильное приложение (app): 1800.00
- Внешние площадки (marketplace): 13775.00
- Веб-сайт (site): 1816.67

Финансовые показатели маркетплейсов превосходят альтернативные каналы почти в 7 раз. Данный феномен вызван исключительно присутствием аномальной транзакции на 45 000 рублей в этой группе. Без учета данного фактора средний чек на маркетплейсах находился бы в районе 3 367 рублей.


In [12]:
# 7.3
channel_stats = channel_stats.sort_values(ascending=False)
print(channel_stats)

channel
marketplace    13775.000000
site            1816.666667
app             1800.000000
Name: order_sum, dtype: float64


## Этап 7.3 — Сортировка результата

Применив метод `sort_values(ascending=False)`, мы упорядочили торговые каналы по весу их среднего чека от большего к меньшему:
1. Площадки marketplace — 13 775.00
2. Официальный веб-ресурс site — 1 816.67
3. Мобильный клиент app — 1 800.00


## Этап 7.4 — Комплексная аналитическая интерпретация

1. Средние чеки по источникам поступления: marketplace лидирует с 13775.00, тогда как site (1816.67) и app (1800.00) демонстрируют схожие скромные результаты.
2. Объем маржинальных операций: зафиксировано 4 крупных чека с флагом `is_high_value = True`.
3. Состояние информации: выявлено 3 пробела в логах времени обслуживания клиентов.
4. Специфика аномалий: единичный заказ на 45000 рублей искажает метрику среднего чека для маркетплейсов. В результате общее среднее по компании (5000) оторвано от медианы (2100).
5. Параметры времени: показатели среднего (39.58) и медианы (37.5) для логистики практически совпадают, демонстрируя отсутствие критических сбоев.

**Резюме:**  
Финансовая структура продаж сильно деформирована из-за одной сверхкрупной сделки на маркетплейсе, которая маскирует реальную ситуацию в каналах site и app. При аудите чеков правильнее ориентироваться на медиану (2100). В то же время показатели скорости доставки выглядят монолитными и предсказуемыми. Около трети базы составляют крупные заказы, а 20% логов имеют пробелы в фиксации времени.


# 8. Итоговый аналитический вывод

Текущий аудит охватывает 15 выполненных заказов. В бизнес-метриках обнаружено 3 пробела в фиксации времени доставки, что требует проверки стабильности работы систем логистического учета. Скорость исполнения заказов находится на стабильном уровне (медиана — 37.5 минут, среднее — 39.58 минут).

Структура выручки по направлениям выглядит следующим образом: маркетплейсы приносят в среднем 13 775 руб. на заказ, сайт — 1 816.67 руб., приложение — 1 800 руб. Однако столь высокая доходность маркетплейсов обманчива: она сформирована всего одной крупной сделкой на 45 000 руб. Большинство клиентов совершают покупки в пределах 2 100 руб. (медианное значение), и лишь 4 транзакции из 15 превышают сумму в 3 000 руб.
